File: Engine/src/tungsten/repl.py
from __future__ import annotations

import sys
from typing import TextIO

from . import __version__
from .expression import EvaluationSession
from .expression import Expr
from .expression import Symbol
from .expression import TungstenExitRequested
from .expression import WolframEvaluationError
from .expression import WolframSyntaxError
from .expression import display_output_parts
from .expression import evaluate
from .expression import parse_input_form


def _banner() -> str:
    return (
        f"Tungsten {__version__} Kernel-free Wolfram Language Interpreter for Microsoft Windows (64-bit)\n"
        "Copyright 2026 OpenAI Codex. Structural subset; not a Wolfram kernel.\n"
    )


def _format_output(expr: Expr) -> str:
    return display_output_parts(expr)[1]


def _output_label(line: int, expr: Expr) -> str:
    form_name, _text = display_output_parts(expr)
    if form_name is None:
        return f"Out[{line}]"
    return f"Out[{line}]//{form_name}"


def _should_print_output(expr: Expr) -> bool:
    return not (isinstance(expr, Symbol) and expr.name == "Null")


def run_repl(
    *,
    stdin: TextIO | None = None,
    stdout: TextIO | None = None,
    stderr: TextIO | None = None,
    show_banner: bool = True,
) -> int:
    input_stream = stdin or sys.stdin
    output_stream = stdout or sys.stdout
    error_stream = stderr or sys.stderr
    session = EvaluationSession()

    if show_banner:
        output_stream.write(_banner())
        output_stream.write("\n")

    while True:
        output_stream.write(f"In[{session.line + 1}]:= ")
        output_stream.flush()

        source = input_stream.readline()
        if source == "":
            output_stream.write("\n")
            output_stream.flush()
            return 0

        source = source.rstrip("\r\n")
        if not source.strip():
            output_stream.write("\n")
            continue

        try:
            parsed = parse_input_form(source)
        except WolframSyntaxError as exc:
            error_stream.write(f"Syntax::sntxi: {exc}\n\n")
            error_stream.flush()
            continue

        line = session.begin_input(source, parsed)
        try:
            result = evaluate(parsed, session=session)
        except TungstenExitRequested as exc:
            return exc.code
        except WolframEvaluationError as exc:
            error_stream.write(f"Evaluate::error: {exc}\n\n")
            error_stream.flush()
            continue

        assert session.current_visible_messages is not None
        assert session.current_prints is not None
        for message in session.current_visible_messages:
            error_stream.write(message.text)
            error_stream.write("\n")
        error_stream.flush()
        for text in session.current_prints:
            output_stream.write(text)
            output_stream.write("\n")

        session.finish_output(result)
        if _should_print_output(result):
            output_stream.write(f"\n{_output_label(line, result)}= {_format_output(result)}\n\n")
        else:
            output_stream.write("\n")
        output_stream.flush()

File: Engine/haskell/src/Tungsten/Repl.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}

-- | A small @wolfram.exe@-style interactive loop over immutable sessions.
module Tungsten.Repl
  ( ReplState (..)
  , ReplStep (..)
  , initialReplState
  , evaluateReplLine
  , runRepl
  ) where

import qualified Data.Map.Strict as Map
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TextIO
import System.IO (hFlush, isEOF, stderr, stdout)
import Tungsten.Evaluate (evaluationErrorMessage)
import Tungsten.Expression
import Tungsten.Parser
import Tungsten.Session

data ReplState = ReplState
  { replSession :: !EvaluationSession
  , replNextLine :: !Integer
  , replInputHistory :: !(Map.Map Integer Expr)
  , replInputStrings :: !(Map.Map Integer Text)
  , replOutputHistory :: !(Map.Map Integer Expr)
  , replMessageHistory :: !(Map.Map Integer [EvaluationMessage])
  }
  deriving (Eq, Show)

data ReplStep
  = ReplValue !Integer !Expr !ReplState
  | ReplExit !Int !ReplState
  | ReplFailure !Text !ReplState
  | ReplEmpty !ReplState
  deriving (Eq, Show)

initialReplState :: ReplState
initialReplState =
  ReplState
    { replSession = emptySession
    , replNextLine = 1
    , replInputHistory = Map.empty
    , replInputStrings = Map.empty
    , replOutputHistory = Map.empty
    , replMessageHistory = Map.empty
    }

evaluateReplLine :: ReplState -> Text -> IO ReplStep
evaluateReplLine state source
  | T.null (T.strip source) = pure (ReplEmpty state)
  | otherwise = case parseInputForm source of
      Left parseError -> pure (ReplFailure (parseErrorMessage parseError) transientState)
      Right parsed -> do
        let line = replNextLine state
            withUnprunedInput =
              transientState
                { replInputHistory = Map.insert line parsed (replInputHistory state)
                , replInputStrings = Map.insert line source (replInputStrings state)
                }
            withInput =
              pruneReplHistory
                line
                transientSession
                withUnprunedInput
            activeSession = installReplHistory line withInput transientSession
            activeState = withInput {replSession = activeSession}
        evaluateInSession activeSession parsed >>= \case
          Left evaluationError ->
            pure
              ( ReplFailure
                  (evaluationErrorMessage evaluationError)
                  activeState {replNextLine = line + 1}
              )
          Right (result, updatedSession) -> pure $ case exitCode result of
            Just code ->
              let finished =
                    finishReplHistory
                      line
                      Nothing
                      activeState
                      updatedSession
               in ReplExit code finished
            Nothing ->
              ReplValue
                line
                result
                (finishReplHistory line (Just result) activeState updatedSession)
 where
  transientSession =
    (replSession state)
      { sessionGeneratedMessages = []
      , sessionVisibleMessages = []
      , sessionPrints = []
      }
  transientState = state {replSession = transientSession}

installReplHistory
  :: Integer
  -> ReplState
  -> EvaluationSession
  -> EvaluationSession
installReplHistory line state session =
  session
    { sessionHistoryLine = Just line
    , sessionInputHistory = replInputHistory state
    , sessionInputStringHistory = replInputStrings state
    , sessionOutputHistory = replOutputHistory state
    , sessionMessageHistory = replMessageHistory state
    }

finishReplHistory
  :: Integer
  -> Maybe Expr
  -> ReplState
  -> EvaluationSession
  -> ReplState
finishReplHistory line output state session =
  let withEffects =
        state
          { replNextLine = line + 1
          , replOutputHistory =
              maybe
                (replOutputHistory state)
                (\value -> Map.insert line value (replOutputHistory state))
                output
          , replMessageHistory =
              Map.insert
                line
                (sessionVisibleMessages session)
                (replMessageHistory state)
          }
      retained = pruneReplHistory line session withEffects
   in retained
        { replSession = installReplHistory line retained session
        }

pruneReplHistory :: Integer -> EvaluationSession -> ReplState -> ReplState
pruneReplHistory line session state =
  case sessionHistoryLengthLimit session of
    Nothing -> state
    Just retainedLength ->
      let cutoff = line - retainedLength + 1
          retained = Map.filterWithKey (\index _ -> index >= cutoff)
       in state
            { replInputHistory = retained (replInputHistory state)
            , replInputStrings = retained (replInputStrings state)
            , replOutputHistory = retained (replOutputHistory state)
            , replMessageHistory = retained (replMessageHistory state)
            }

exitCode :: Expr -> Maybe Int
exitCode = \case
  Symbol "Exit" -> Just 0
  Symbol "Quit" -> Just 0
  Call (Symbol "Exit") [] -> Just 0
  Call (Symbol "Quit") [] -> Just 0
  Call (Symbol "Exit") [Integer value] -> Just (fromInteger value)
  Call (Symbol "Quit") [Integer value] -> Just (fromInteger value)
  _ -> Nothing

runRepl :: Bool -> IO Int
runRepl showBanner = do
  if showBanner
    then TextIO.putStrLn "Tungsten Haskell kernel-free REPL. Exit with Exit[] or Quit[]."
    else pure ()
  loop initialReplState
 where
  loop state = do
    TextIO.putStr ("In[" <> T.pack (show (replNextLine state)) <> "]:= ")
    hFlush stdout
    finished <- isEOF
    if finished
      then TextIO.putStrLn "" *> pure 0
      else do
        source <- TextIO.getLine
        evaluateReplLine state source >>= \case
          ReplEmpty updated -> loop updated
          ReplFailure message updated -> TextIO.putStrLn ("Error: " <> message) *> loop updated
          ReplExit code updated -> emitSessionStreams updated *> pure code
          ReplValue line value updated -> do
            emitSessionStreams updated
            TextIO.putStrLn ("Out[" <> T.pack (show line) <> "]= " <> fullForm value)
            loop updated

  emitSessionStreams updated = do
    mapM_ TextIO.putStrLn (sessionPrints (replSession updated))
    mapM_
      (TextIO.hPutStrLn stderr . evaluationMessageText)
      (sessionVisibleMessages (replSession updated))
    hFlush stderr
